# Customer Behavior Analytics
## Business Problem Statement

A leading retail company wants to better understand customer shopping behavior to improve sales, customer satisfaction, and long-term loyalty.

This project analyzes customer purchase behavior using Python, PostgreSQL, SQL, and Power BI to identify customer trends, purchasing patterns, revenue drivers, and actionable business insights.

### End-to-End Retail Analytics Project
**Objective:**
Analyze customer shopping behavior to identify purchasing trends, customer segments, and business insights that support data-driven decision-making using Python, PostgreSQL, SQL, and Power BI.

**Tools & Technologies:**
- Python (Pandas)
- PostgreSQL
- SQL
- Power BI
- Git & GitHub

## 1. Import Libraries

In [2]:
import pandas as pd

## 2. Load the Dataset

In [3]:
# Load the customer shopping behavior dataset
df = pd.read_csv("../Data/customer_shopping_behavior.csv")

# Preview the first five rows
df.head()

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,3.1,Yes,Express,Yes,Yes,14,Venmo,Fortnightly
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,3.1,Yes,Express,Yes,Yes,2,Cash,Fortnightly
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,3.1,Yes,Free Shipping,Yes,Yes,23,Credit Card,Weekly
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,3.5,Yes,Next Day Air,Yes,Yes,49,PayPal,Weekly
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,2.7,Yes,Free Shipping,Yes,Yes,31,PayPal,Annually


### 3. Exploratory Data Analysis (EDA)

In [5]:
# Display dataset dimensions
df.shape

(3900, 18)

In [6]:
# Display column names
df.columns

Index(['Customer ID', 'Age', 'Gender', 'Item Purchased', 'Category',
       'Purchase Amount (USD)', 'Location', 'Size', 'Color', 'Season',
       'Review Rating', 'Subscription Status', 'Shipping Type',
       'Discount Applied', 'Promo Code Used', 'Previous Purchases',
       'Payment Method', 'Frequency of Purchases'],
      dtype='str')

In [7]:
# Display data types and missing values
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer ID             3900 non-null   int64  
 1   Age                     3900 non-null   int64  
 2   Gender                  3900 non-null   str    
 3   Item Purchased          3900 non-null   str    
 4   Category                3900 non-null   str    
 5   Purchase Amount (USD)   3900 non-null   int64  
 6   Location                3900 non-null   str    
 7   Size                    3900 non-null   str    
 8   Color                   3900 non-null   str    
 9   Season                  3900 non-null   str    
 10  Review Rating           3863 non-null   float64
 11  Subscription Status     3900 non-null   str    
 12  Shipping Type           3900 non-null   str    
 13  Discount Applied        3900 non-null   str    
 14  Promo Code Used         3900 non-null   str    
 15

In [8]:
# Summary statistics
df.describe()

,Customer ID,Age,Purchase Amount (USD),Review Rating,Previous Purchases
count,3900.000000,3900.000000,3900.000000,3863.000000,3900.000000
mean,1950.500000,44.068462,59.764359,3.750065,25.351538
std,1125.977353,15.207589,23.685392,0.716983,14.447125
min,1.000000,18.000000,20.000000,2.500000,1.000000
25%,975.750000,31.000000,39.000000,3.100000,13.000000
50%,1950.500000,44.000000,60.000000,3.800000,25.000000
75%,2925.250000,57.000000,81.000000,4.400000,38.000000
max,3900.000000,70.000000,100.000000,5.000000,50.000000


### Key Observations

- The dataset contains **3,900 customer records** and **18 features**.
- Customer age ranges from **18 to 70 years**.
- The average purchase amount is approximately **$59.76**.
- Only the **Review Rating** column contains missing values (37 records).
- Most variables are categorical, making customer segmentation possible.

## 4. Data Cleaning & Preprocessing

In [9]:
# Check missing values
df.isnull().sum()

Customer ID                0
Age                        0
Gender                     0
Item Purchased             0
Category                   0
Purchase Amount (USD)      0
Location                   0
Size                       0
Color                      0
Season                     0
Review Rating             37
Subscription Status        0
Shipping Type              0
Discount Applied           0
Promo Code Used            0
Previous Purchases         0
Payment Method             0
Frequency of Purchases     0
dtype: int64

### Observations

- The dataset contains missing values only in the **Review Rating** column.
- A total of **37 records** (less than 1% of the dataset) have missing review ratings.
- Since the missing proportion is very small, the values can be imputed rather than removing entire records.

### Handling Missing Values

The `Review Rating` column contains 37 missing values. Instead of replacing them with the overall median, the missing values will be imputed using the median review rating within each product category. This approach preserves category-specific rating patterns and reduces potential bias.

In [10]:
# Fill missing review ratings using the median rating within each category
df["Review Rating"] = (
    df.groupby("Category")["Review Rating"]
      .transform(lambda x: x.fillna(x.median()))
)

In [11]:
# Verify that all missing values have been handled
df.isnull().sum()

Customer ID               0
Age                       0
Gender                    0
Item Purchased            0
Category                  0
Purchase Amount (USD)     0
Location                  0
Size                      0
Color                     0
Season                    0
Review Rating             0
Subscription Status       0
Shipping Type             0
Discount Applied          0
Promo Code Used           0
Previous Purchases        0
Payment Method            0
Frequency of Purchases    0
dtype: int64

### Result

All missing values in the dataset have been successfully handled. The Review Rating column was imputed using the median rating within each product category, preserving category-level characteristics while maintaining all customer records.

## 5. Standardize Column Names

To improve code readability and maintain consistency across Python, SQL, and Power BI, all column names will be converted to lowercase and spaces will be replaced with underscores (snake_case naming convention).

In [12]:
# Convert column names to lowercase and replace spaces with underscores
df.columns = (
    df.columns
      .str.lower()
      .str.replace(" ", "_")
)

In [13]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount_(usd)', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'promo_code_used', 'previous_purchases',
       'payment_method', 'frequency_of_purchases'],
      dtype='str')

In [14]:
# Rename purchase_amount_(usd) to purchase_amount
df = df.rename(
    columns={
        "purchase_amount_(usd)": "purchase_amount"
    }
)

In [15]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'promo_code_used', 'previous_purchases',
       'payment_method', 'frequency_of_purchases'],
      dtype='str')

## 5. Feature Engineering

Create new variables that improve business analysis.

In [16]:
# Create age groups

df["age_group"] = pd.cut(
    df["age"],
    bins=[17, 25, 35, 50, 100],
    labels=["18-25", "26-35", "36-50", "51+"]
)

In [17]:
df[["age", "age_group"]].head(10)

,age,age_group
0,55,51+
1,19,18-25
2,50,36-50
3,21,18-25
4,45,36-50
5,46,36-50
6,63,51+
7,27,26-35
8,26,26-35
9,57,51+


## 8. Export the Cleaned Dataset

Save the cleaned and transformed dataset for SQL analysis and Tableau visualization.

In [54]:
# Export cleaned dataset

df.to_csv(
    "../Data/customer_shopping_behavior_cleaned.csv",
    index=False
)

print("Cleaned dataset exported successfully!")

Cleaned dataset exported successfully!


## 9. Load Cleaned Data into PostgreSQL

In [ ]:
# Import the libraries
from sqlalchemy import create_engine

In [56]:
# Create the connection
engine = create_engine(
    "postgresql+psycopg2://postgres:Yomari@localhost:5432/customer_behavior_db"
)

In [59]:
#Upload the dataframe
df.to_sql(
    "customer_behavior",
    engine,
    if_exists="replace",
    index=False
)
print("Data uploaded successfully!")

Data uploaded successfully!


## 10. SQL Business Analysis (PostgreSQL)

The cleaned customer behavior dataset was imported into PostgreSQL for business analysis using SQL. The following queries answer key business questions related to customer spending, purchasing behavior, demographics, and product performance.